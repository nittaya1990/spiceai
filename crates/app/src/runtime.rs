/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::{fmt::Display, str::FromStr, sync::Arc};

use spicepod::component::runtime::UserAgentCollection;

use super::App;

impl App {
    /// Get a parameter from the app's runtime params, with a default value if the parameter is not set or is not valid.
    ///
    /// Returns `default_value` if the parameter is not set or is not valid.
    ///
    /// If the parameter is set but is not valid, logs a warning and returns `default_value`.
    #[must_use]
    pub fn get_runtime_param<T>(app: &Option<Arc<Self>>, param: &str, default_value: T) -> T
    where
        T: Display + FromStr,
    {
        let Some(value) = app.as_ref().and_then(|app| app.runtime.params.get(param)) else {
            return default_value;
        };

        if let Ok(parsed_value) = value.parse::<T>() {
            parsed_value
        } else {
            eprintln!("runtime.params.{param} is not valid, defaulting to {default_value}");
            default_value
        }
    }

    #[must_use]
    pub fn user_agent_collection(&self) -> UserAgentCollection {
        self.runtime.telemetry.user_agent_collection
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::AppBuilder;
    use std::{collections::HashMap, sync::Arc};

    fn create_app_with_params(params: HashMap<String, String>) -> Arc<App> {
        Arc::new(AppBuilder::new("test").with_runtime_params(params).build())
    }

    #[test]
    fn test_get_runtime_param() {
        // Test case 1: Parameter is not set
        let app = Some(create_app_with_params(HashMap::new()));
        assert!(App::get_runtime_param(&app, "test_param", true));
        assert!(!App::get_runtime_param(&app, "test_param", false));

        // Test case 2: Parameter is set to "true"
        let mut params = HashMap::new();
        params.insert("test_param".to_string(), "true".to_string());
        let app = Some(create_app_with_params(params));
        assert!(App::get_runtime_param(&app, "test_param", false));

        // Test case 3: Parameter is set to "false"
        let mut params = HashMap::new();
        params.insert("test_param".to_string(), "false".to_string());
        let app = Some(create_app_with_params(params));
        assert!(!App::get_runtime_param(&app, "test_param", true));

        // Test case 4: Parameter is set to an invalid boolean value
        let mut params = HashMap::new();
        params.insert("test_param".to_string(), "not_a_bool".to_string());
        let app = Some(create_app_with_params(params));
        assert!(App::get_runtime_param(&app, "test_param", true));
        assert!(!App::get_runtime_param(&app, "test_param", false));

        // Test case 5: App is None
        assert!(App::get_runtime_param(&None, "test_param", true));
        assert!(!App::get_runtime_param(&None, "test_param", false));
    }
}
