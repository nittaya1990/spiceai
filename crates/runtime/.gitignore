abc-sqlite.db-wal
abc-sqlite.db-shm
