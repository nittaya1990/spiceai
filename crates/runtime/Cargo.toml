[package]
description = "The spiced runtime"
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "runtime"
repository.workspace = true
rust-version.workspace = true
version.workspace = true

[dependencies]
app = { path = "../app" }
arrow.workspace = true
arrow-csv = "53.3.0"
arrow-flight = { workspace = true, features = ["flight-sql-experimental"] }
arrow-ipc.workspace = true
arrow-json.workspace = true
arrow-schema = { workspace = true, features = ["serde"] }
arrow_sql_gen = { path = "../arrow_sql_gen" }
arrow_tools = { path = "../arrow_tools" }
async-openai.workspace = true
async-stream.workspace = true
async-trait.workspace = true
aws-config = { version = "1.1.10", optional = true }
aws-sdk-dynamodb = { version = "1", optional = true }
aws-sdk-secretsmanager = { version = "1", optional = true }
aws-sdk-sts = { version = "1", optional = true }
axum.workspace = true
axum-extra = { version = "0.9.6", features = ["typed-header"] }
base64.workspace = true
bytes.workspace = true
cache = { path = "../cache" }
chrono = { version = "0.4.38" }
clap.workspace = true
clickhouse-rs = { workspace = true, optional = true }
csv = "1.3.0"
dashmap = "6.1.0"
data_components = { path = "../data_components" }
datafusion.workspace = true
datafusion-federation = { workspace = true }
datafusion-federation-sql = { workspace = true }
datafusion-functions-json = { workspace = true }
datafusion-table-providers = { workspace = true }
db_connection_pool = { path = "../db_connection_pool" }
document_parse = { path = "../document_parse" }
dotenvy.workspace = true
duckdb = { workspace = true, features = [
    "bundled",
    "r2d2",
    "vtab",
    "vtab-arrow",
], optional = true }
flight_client = { path = "../flight_client" }
fundu = { workspace = true }
futures.workspace = true
globset.workspace = true
governor = "0.8.0"
graph-rs-sdk = { workspace = true, optional = true }
graphql-parser.workspace = true
headers-accept = "0.1.3"
http.workspace = true
http-body-util = "0.1.2"
hyper = "1.6.0"
hyper-util = { version = "0.1.6", features = ["service"] }
iceberg.workspace = true
iceberg-catalog-rest.workspace = true
imap = { workspace = true, optional = true }
indexmap.workspace = true
itertools.workspace = true
jsonwebtoken = "9.3.0"
keyring = { version = "3.6.1", features = [
    "apple-native",
    "windows-native",
    "linux-native",
], optional = true }
llms = { path = "../llms" }
logos = "0.15.0"
mcp-client = { git = "https://github.com/block/goose", package = "mcp-client", optional = true }
mcp-core = { git = "https://github.com/block/goose", package = "mcp-core", optional = true }
mediatype = "0.19.18"
model_components = { path = "../model_components" }
mysql_async = { workspace = true, optional = true }
notify = "7.0.0"
ns_lookup = { path = "../ns_lookup" }
object_store = { workspace = true, features = ["aws", "http", "azure"] }
once_cell = "1.19.0"
opentelemetry.workspace = true
opentelemetry-prometheus.workspace = true
opentelemetry-proto = { version = "0.27", features = [
    "gen-tonic-messages",
    "gen-tonic",
    "metrics",
] }
opentelemetry_sdk.workspace = true
otel-arrow = { path = "../otel-arrow" }
package = { path = "../package" }
percent-encoding.workspace = true
pin-project.workspace = true
prometheus.workspace = true
prometheus-parse = "0.2.5"
prost = { version = "0.13.1", default-features = false, features = [
    "prost-derive",
] }
rand = "0.8.5"
regex.workspace = true
reqwest.workspace = true
runtime-auth = { path = "../runtime-auth" }
rusqlite = { workspace = true, optional = true }
rustls.workspace = true
rustls-pemfile.workspace = true
schemars = "0.8.19"
secrecy.workspace = true
serde = { workspace = true, features = ["derive"] }
serde_json.workspace = true
snafu.workspace = true
snowflake-api = { workspace = true, optional = true }
spicepod = { path = "../spicepod" }
ssh2 = { workspace = true, optional = true }
suppaftp = { workspace = true, optional = true }
telemetry = { path = "../telemetry" }
tiberius = { workspace = true, optional = true }
tokio.workspace = true
tokio-rusqlite = { workspace = true, optional = true }
tokio-rustls = "0.26.0"
tokio-stream.workspace = true
tokio-util = { workspace = true, optional = true }
tonic.workspace = true
tonic-health.workspace = true
tower.workspace = true
tower-http.workspace = true
tracing.workspace = true
tracing-futures.workspace = true
tract-core = "0.21.0"
url = "2.5.4"
util = { path = "../util" }
utoipa = { version = "5", features = ["axum_extras"], optional = true }
utoipa-swagger-ui = { version = "8", features = ["axum"], optional = true }

uuid = { workspace = true, features = ["v7"] }
x509-certificate.workspace = true

[dev-dependencies]
anyhow = "1.0.86"
async-graphql = "7.0.5"
async-graphql-axum = "=7.0.13"
azure_core = "0.21.0"
azure_storage = "0.21.0"
azure_storage_blobs = "0.21.0"
bollard = "0.18.1"
dirs = "5.0.1"
flightrepl = { path = "../flightrepl" }
hf-hub = { workspace = true }
insta.workspace = true
jsonpath-rust = "0.7.3"
opentelemetry_sdk = { workspace = true, features = ["rt-tokio-current-thread"] }
paste.workspace = true
scopeguard = "1.2.0"
serde_json = { workspace = true, features = ["preserve_order"] }
spice-cloud = { path = "../spice_cloud" }
test-framework = { path = "../test-framework" }
tokio = { workspace = true, features = ["time", "test-util"] }
tracing-opentelemetry.workspace = true
tracing-subscriber.workspace = true

[features]
aws-secrets-manager = [
    "dep:aws-config",
    "dep:aws-sdk-secretsmanager",
    "dep:aws-sdk-sts",
]
clickhouse = [
    "dep:clickhouse-rs",
    "db_connection_pool/clickhouse",
    "arrow_sql_gen/clickhouse",
    "data_components/clickhouse",
]
cuda = ["llms/cuda"]
databricks = ["data_components/databricks"]
debezium = ["data_components/debezium"]
default = ["keyring-secret-store", "aws-secrets-manager", "sharepoint"]
delta_lake = ["data_components/delta_lake"]
dev = ["openapi", "otel-arrow/dev", "mcp"]
dremio = []
duckdb = ["dep:duckdb", "db_connection_pool/duckdb", "data_components/duckdb"]
dynamodb = ["dep:aws-sdk-dynamodb", "data_components/dynamodb"]
flightsql = ["data_components/flightsql"]
ftp = ["dep:suppaftp", "dep:ssh2"]
imap = ["data_components/imap", "dep:imap"]
keyring-secret-store = ["dep:keyring"]
mcp = ["dep:mcp-client", "dep:mcp-core"]
metal = ["llms/metal"]
models = ["model_components/full"]
mssql = ["dep:tiberius", "dep:tokio-util", "data_components/mssql"]
mysql = ["dep:mysql_async", "db_connection_pool/mysql", "data_components/mysql"]
odbc = ["db_connection_pool/odbc", "data_components/odbc"]
openapi = ["dep:utoipa-swagger-ui", "dep:utoipa"]
postgres = [
    "db_connection_pool/postgres",
    "data_components/postgres",
    "datafusion-table-providers/postgres",
]
postgres-write = []
sharepoint = ["data_components/sharepoint", "dep:graph-rs-sdk"]
snowflake = [
    "dep:snowflake-api",
    "db_connection_pool/snowflake",
    "data_components/snowflake",
]
spark = ["data_components/spark_connect"]
sqlite = [
    "dep:tokio-rusqlite",
    "db_connection_pool/sqlite",
    "data_components/sqlite",
    "dep:rusqlite",
]

[[bench]]
harness = false
name = "bench"

[[bench]]
harness = false
name = "vector_search"

[target.'cfg(windows)'.dependencies]
winver = "1.0.0"
