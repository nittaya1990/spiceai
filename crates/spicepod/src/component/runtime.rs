/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::{collections::HashMap, error::Error, sync::Arc};

use super::is_default;
#[cfg(feature = "schemars")]
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

const TASK_HISTORY_RETENTION_MINIMUM: u64 = 60; // 1 minute

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
#[cfg_attr(feature = "schemars", derive(JsonSchema))]
#[serde(deny_unknown_fields)]
pub struct Runtime {
    #[serde(default, skip_serializing_if = "is_default")]
    pub results_cache: ResultsCache,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub dataset_load_parallelism: Option<usize>,

    /// If set, the runtime will configure all endpoints to use TLS
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tls: Option<TlsConfig>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub tracing: Option<TracingConfig>,

    #[serde(default, skip_serializing_if = "is_default")]
    pub telemetry: TelemetryConfig,

    #[serde(skip_serializing_if = "HashMap::is_empty")]
    #[serde(default)]
    pub params: HashMap<String, String>,

    #[serde(default, skip_serializing_if = "is_default")]
    pub task_history: TaskHistory,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub auth: Option<Auth>,

    #[serde(default, skip_serializing_if = "is_default")]
    pub cors: CorsConfig,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[cfg_attr(feature = "schemars", derive(JsonSchema))]
pub struct ResultsCache {
    #[serde(default = "default_true")]
    pub enabled: bool,
    pub cache_max_size: Option<String>,
    pub item_ttl: Option<String>,
    pub eviction_policy: Option<String>,
}

const fn default_true() -> bool {
    true
}

impl Default for ResultsCache {
    fn default() -> Self {
        Self {
            enabled: true,
            cache_max_size: None,
            item_ttl: None,
            eviction_policy: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
#[cfg_attr(feature = "schemars", derive(JsonSchema))]
pub struct TlsConfig {
    /// If set, the runtime will configure all endpoints to use TLS
    pub enabled: bool,

    /// A filesystem path to a file containing the PEM encoded certificate
    pub certificate_file: Option<String>,

    /// A PEM encoded certificate
    pub certificate: Option<String>,

    /// A filesystem path to a file containing the PEM encoded private key
    pub key_file: Option<String>,

    /// A PEM encoded private key
    pub key: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
#[cfg_attr(feature = "schemars", derive(JsonSchema))]
pub struct TracingConfig {
    pub zipkin_enabled: bool,
    pub zipkin_endpoint: Option<String>,
}

#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq, Default)]
#[serde(deny_unknown_fields)]
#[cfg_attr(feature = "schemars", derive(JsonSchema))]
#[serde(rename_all = "camelCase")]
pub enum UserAgentCollection {
    #[default]
    Full,
    Disabled,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
#[cfg_attr(feature = "schemars", derive(JsonSchema))]
pub struct TelemetryConfig {
    #[serde(default = "default_true")]
    pub enabled: bool,
    #[serde(default)]
    pub user_agent_collection: UserAgentCollection,
    #[serde(default)]
    pub properties: HashMap<String, String>,
}

impl Default for TelemetryConfig {
    fn default() -> Self {
        Self {
            enabled: true,
            user_agent_collection: UserAgentCollection::default(),
            properties: HashMap::new(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
#[cfg_attr(feature = "schemars", derive(JsonSchema))]
pub struct TaskHistory {
    #[serde(default = "default_true")]
    pub enabled: bool,
    #[serde(default = "default_none")]
    pub captured_output: Arc<str>,
    #[serde(default = "default_retention_period")]
    pub retention_period: Arc<str>,
    #[serde(default = "default_retention_check_interval")]
    pub retention_check_interval: Arc<str>,
}

fn default_none() -> Arc<str> {
    "none".into()
}

fn default_retention_period() -> Arc<str> {
    "8h".into()
}

fn default_retention_check_interval() -> Arc<str> {
    "15m".into()
}

impl Default for TaskHistory {
    fn default() -> Self {
        Self {
            enabled: true,
            captured_output: default_none(),
            retention_period: default_retention_period(),
            retention_check_interval: default_retention_check_interval(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Default)]
pub enum TaskHistoryCapturedOutput {
    #[default]
    None,
    Truncated,
}

impl TaskHistory {
    pub fn get_captured_output(
        &self,
    ) -> Result<TaskHistoryCapturedOutput, Box<dyn Error + Send + Sync>> {
        if self.captured_output == "none".into() {
            return Ok(TaskHistoryCapturedOutput::None);
        } else if self.captured_output == "truncated".into() {
            return Ok(TaskHistoryCapturedOutput::Truncated);
        }

        Err(format!(
            r#"Expected "none" or "truncated" for "captured_output", but got: "{}""#,
            self.captured_output
        )
        .into())
    }

    fn retention_value_as_secs(
        value: &str,
        field: &str,
    ) -> Result<u64, Box<dyn Error + Send + Sync>> {
        let duration = fundu::parse_duration(value).map_err(|e| e.to_string())?;

        if duration.as_secs() < TASK_HISTORY_RETENTION_MINIMUM {
            return Err(format!(
                r#"Task history retention {field} must be at least {TASK_HISTORY_RETENTION_MINIMUM} seconds. To disable task history, set the property "enabled: false"."#,
            ).into());
        }

        Ok(duration.as_secs())
    }

    pub fn retention_period_as_secs(&self) -> Result<u64, Box<dyn Error + Send + Sync>> {
        Self::retention_value_as_secs(&self.retention_period, "period")
    }

    pub fn retention_check_interval_as_secs(&self) -> Result<u64, Box<dyn Error + Send + Sync>> {
        Self::retention_value_as_secs(&self.retention_check_interval, "check interval")
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[cfg_attr(feature = "schemars", derive(JsonSchema))]
#[serde(deny_unknown_fields)]
pub struct Auth {
    #[serde(alias = "api-key")]
    pub api_key: Option<ApiKeyAuth>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[cfg_attr(feature = "schemars", derive(JsonSchema))]
#[serde(deny_unknown_fields)]
pub struct ApiKeyAuth {
    #[serde(default = "default_true")]
    pub enabled: bool,
    pub keys: Vec<ApiKey>,
}

#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(feature = "schemars", derive(JsonSchema))]
pub enum ApiKey {
    ReadOnly { key: String },
    ReadWrite { key: String },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[cfg_attr(feature = "schemars", derive(JsonSchema))]
#[serde(deny_unknown_fields)]
pub struct CorsConfig {
    #[serde(default)]
    pub enabled: bool,
    #[serde(default = "default_allowed_origins")]
    pub allowed_origins: Vec<String>,
}

fn default_allowed_origins() -> Vec<String> {
    vec!["*".to_string()]
}

impl Default for CorsConfig {
    fn default() -> Self {
        Self {
            enabled: false,
            allowed_origins: default_allowed_origins(),
        }
    }
}

impl ApiKey {
    #[must_use]
    pub fn parse_str(input: &str) -> Self {
        if let Some((key, kind)) = input.rsplit_once(':') {
            match kind {
                "ro" => ApiKey::ReadOnly {
                    key: key.to_string(),
                },
                "rw" => ApiKey::ReadWrite {
                    key: key.to_string(),
                },
                _ => ApiKey::ReadOnly {
                    key: input.to_string(),
                },
            }
        } else {
            // Default to ReadOnly if no suffix is provided
            ApiKey::ReadOnly {
                key: input.to_string(),
            }
        }
    }
}

impl<'de> Deserialize<'de> for ApiKey {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let input = String::deserialize(deserializer)?;

        Ok(ApiKey::parse_str(&input))
    }
}

impl Serialize for ApiKey {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match self {
            ApiKey::ReadOnly { key } => serializer.serialize_str(key),
            ApiKey::ReadWrite { key } => serializer.serialize_str(&format!("{key}:rw")),
        }
    }
}

impl PartialEq<str> for ApiKey {
    fn eq(&self, other: &str) -> bool {
        match self {
            ApiKey::ReadOnly { key } | ApiKey::ReadWrite { key } => key == other,
        }
    }
}

impl AsRef<str> for ApiKey {
    fn as_ref(&self) -> &str {
        match self {
            ApiKey::ReadOnly { key } | ApiKey::ReadWrite { key } => key.as_str(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_yaml;

    #[test]
    fn test_deserialize_api_keys() {
        let yaml = r"
        api_key:
            enabled: true
            keys:
                - api-key-1
                - api-key-2:ro
                - api-key-3:rw
        ";

        let parsed: Auth = serde_yaml::from_str(yaml).expect("Failed to parse Auth");

        let api_key = parsed.api_key.expect("api_key section exists");

        assert_eq!(
            api_key.keys[0],
            ApiKey::ReadOnly {
                key: "api-key-1".to_string()
            }
        );
        assert_eq!(
            api_key.keys[1],
            ApiKey::ReadOnly {
                key: "api-key-2".to_string()
            }
        );
        assert_eq!(
            api_key.keys[2],
            ApiKey::ReadWrite {
                key: "api-key-3".to_string()
            }
        );
    }

    #[test]
    fn test_deserialize_api_key_alternative_name() {
        let yaml = r"
        api-key:
            enabled: true
            keys:
                - api-key-1
        ";

        let parsed: Auth = serde_yaml::from_str(yaml).expect("Failed to parse Auth");

        let api_key = parsed.api_key.expect("api_key section exists");

        assert_eq!(
            api_key.keys[0],
            ApiKey::ReadOnly {
                key: "api-key-1".to_string()
            }
        );
    }
}
