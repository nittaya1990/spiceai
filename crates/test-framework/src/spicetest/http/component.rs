/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use anyhow::{anyhow, Result};
use async_openai::{
    config::OpenAIConfig,
    types::{
        ChatCompletionRequestMessage, ChatCompletionRequestUserMessageArgs,
        CreateChatCompletionRequest, CreateChatCompletionRequestArgs, CreateEmbeddingRequest,
        EncodingFormat,
    },
    Client as OpenAIClient,
};
use reqwest::Client;
use std::time::{Duration, Instant};

/// A component within the Spiced instance to test for consistency.
///
/// This component must be accessible over HTTP.
#[derive(Clone)]
pub enum HttpComponent {
    Model {
        model: String,
        api_base: String,
    },
    Embedding {
        embedding: String,
        api_base: String,
    },
    Generic {
        // For a generic component, must know full URL.
        http_url: String,
        component_name: String,
    },
}

impl HttpComponent {
    fn api_base(&self) -> String {
        match self {
            HttpComponent::Generic {
                http_url: api_base, ..
            }
            | HttpComponent::Model { api_base, .. }
            | HttpComponent::Embedding { api_base, .. } => api_base.clone(),
        }
    }

    #[must_use]
    pub fn component_name(&self) -> String {
        match self {
            HttpComponent::Generic {
                component_name: model,
                ..
            }
            | HttpComponent::Model { model, .. } => model.clone(),
            HttpComponent::Embedding { embedding, .. } => embedding.clone(),
        }
    }

    #[must_use]
    pub fn with_api_base(self, api_base: String) -> Self {
        match self {
            HttpComponent::Model { model, .. } => HttpComponent::Model { model, api_base },
            HttpComponent::Embedding { embedding, .. } => HttpComponent::Embedding {
                embedding,
                api_base,
            },
            HttpComponent::Generic { component_name, .. } => HttpComponent::Generic {
                http_url: api_base,
                component_name,
            },
        }
    }

    #[must_use]
    pub fn with_component_name(self, component_name: String) -> Self {
        match self {
            HttpComponent::Model { api_base, .. } => HttpComponent::Model {
                model: component_name,
                api_base,
            },
            HttpComponent::Embedding { api_base, .. } => HttpComponent::Embedding {
                embedding: component_name,
                api_base,
            },
            HttpComponent::Generic {
                http_url: http_base,
                ..
            } => HttpComponent::Generic {
                http_url: http_base,
                component_name,
            },
        }
    }

    /// Sends a request to the component and returns the duration of the request.
    /// Payload may be the entire HTTP request body, or a portion of it (dependent of the component).
    pub async fn send_request(&self, client: &Client, payload: &str) -> Result<Duration> {
        let start_time = Instant::now();
        match self {
            HttpComponent::Generic {
                http_url: http_base,
                ..
            } => {
                let req = client.post(http_base.clone())
                    .body(payload.to_string()).send()
                    .await
                    .map_err(|e| anyhow!("Error received from generic HTTP POST request to {http_base}. Error: {e:?}"))?;

                req.error_for_status()
                    .map_err(|e| anyhow!("Received error status from {http_base}. Error: {e:?}"))?;
            }
            HttpComponent::Model { model, .. } => {
                let c = OpenAIClient::with_config(
                    OpenAIConfig::default().with_api_base(self.api_base()),
                )
                .with_http_client(client.clone())
                .clone();

                let req: CreateChatCompletionRequest =
                    match serde_json::from_str::<CreateChatCompletionRequest>(payload) {
                        Ok(mut req) => {
                            // Ensure the model is overriden.
                            req.model.clone_from(model);
                            req
                        }
                        Err(_) => CreateChatCompletionRequestArgs::default()
                            .model(model.clone())
                            .messages(vec![ChatCompletionRequestMessage::User(
                                ChatCompletionRequestUserMessageArgs::default()
                                    .content(payload.to_string())
                                    .build()
                                    .map_err(|e| {
                                        anyhow!("failed to build user message. Error: {e:?}")
                                    })?,
                            )])
                            .build()
                            .map_err(|e| anyhow!("Failed to build model request. Error: {e:?}"))?,
                    };
                let _ = c.chat().create(req).await?;
            }
            HttpComponent::Embedding { embedding, .. } => {
                let c = OpenAIClient::with_config(
                    OpenAIConfig::default().with_api_base(self.api_base()),
                )
                .with_http_client(client.clone())
                .clone();
                let req: CreateEmbeddingRequest =
                    match serde_json::from_str::<CreateEmbeddingRequest>(payload) {
                        Ok(mut req) => {
                            // Ensure the model is overriden.
                            req.model.clone_from(embedding);
                            req
                        }
                        Err(_) => CreateEmbeddingRequest {
                            model: embedding.clone(),
                            input: async_openai::types::EmbeddingInput::String(payload.to_string()),
                            encoding_format: Some(EncodingFormat::Float),
                            user: None,
                            dimensions: None,
                        },
                    };
                c.embeddings().create(req).await?;
            }
        }
        Ok(start_time.elapsed())
    }
}
