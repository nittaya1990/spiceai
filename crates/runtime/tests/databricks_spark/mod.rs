/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::sync::Arc;

use app::AppBuilder;

use crate::{
    get_test_datafusion, init_tracing, run_query_and_check_results, utils::test_request_context,
    ValidateFn,
};
use runtime::{status, Runtime};
use spicepod::component::{catalog::Catalog, params::Params};

fn make_catalog(path: &str, name: &str) -> Catalog {
    let mut catalog = Catalog::new(format!("databricks:{path}"), name.to_string());
    catalog.params = Some(get_params());
    catalog
}

#[allow(clippy::expect_used)]
fn get_params() -> Params {
    // Verify that the environment variables are set
    let _ = std::env::var("DATABRICKS_HOST").expect("DATABRICKS_HOST is not set");
    let _ = std::env::var("DATABRICKS_TOKEN").expect("DATABRICKS_TOKEN is not set");
    let _ = std::env::var("DATABRICKS_CLUSTER_ID").expect("DATABRICKS_CLUSTER_ID is not set");

    Params::from_string_map(
        vec![
            (
                "databricks_endpoint".to_string(),
                "${ env:DATABRICKS_HOST }".to_string(),
            ),
            (
                "databricks_token".to_string(),
                "${ env:DATABRICKS_TOKEN }".to_string(),
            ),
            (
                "databricks_cluster_id".to_string(),
                "${ env:DATABRICKS_CLUSTER_ID }".to_string(),
            ),
            ("mode".to_string(), "spark_connect".to_string()),
        ]
        .into_iter()
        .collect(),
    )
}

#[tokio::test]
async fn databricks_spark_integration_test() -> Result<(), anyhow::Error> {
    type QueryTests<'a> = Vec<(&'a str, &'a str, Option<Box<ValidateFn>>)>;
    let _ = rustls::crypto::CryptoProvider::install_default(
        rustls::crypto::aws_lc_rs::default_provider(),
    );
    let _tracing = init_tracing(Some("integration=debug,info"));

    test_request_context()
        .scope(async {
            let app = AppBuilder::new("databricks_spark_connector")
                .with_catalog(make_catalog(
                    "catalog-dash-test",
                    "db_uc",
                ))
                .build();

            let status = status::RuntimeStatus::new();
            let df = get_test_datafusion(Arc::clone(&status));

            let mut rt = Runtime::builder()
                .with_datafusion(df)
                .with_app(app)
                .build()
                .await;

            // Set a timeout for the test
            tokio::select! {
                () = tokio::time::sleep(std::time::Duration::from_secs(10)) => {
                    return Err(anyhow::anyhow!("Timed out waiting for datasets to load"));
                }
                () = rt.load_components() => {}
            }

            let queries: QueryTests = vec![(
                "SELECT * FROM db_uc.default.databricks_demo ORDER BY network, isp LIMIT 10",
                "select",
                Some(Box::new(|result_batches| {
                    for batch in &result_batches {
                        assert_eq!(batch.num_columns(), 7, "num_cols: {}", batch.num_columns());
                        assert_eq!(batch.num_rows(), 10, "num_rows: {}", batch.num_rows());
                    }

                    // snapshot the values of the results
                    let results = arrow::util::pretty::pretty_format_batches(&result_batches)
                        .expect("should pretty print result batch");
                    insta::with_settings!({
                        description => format!("Databricks (mode: spark_connect) Integration Test Results"),
                        omit_expression => true,
                        snapshot_path => "../snapshots"
                    }, {
                        insta::assert_snapshot!(format!("databricks_spark_connect_select"), results);
                    });
                })),
            )];

            for (query, snapshot_suffix, validate_result) in queries {
                run_query_and_check_results(
                    &mut rt,
                    &format!("databricks_spark_connect_test_{snapshot_suffix}"),
                    query,
                    true,
                    validate_result,
                )
                .await
                .map_err(|e| anyhow::anyhow!("{e}"))?;
            }

            Ok(())
        })
        .await
}
