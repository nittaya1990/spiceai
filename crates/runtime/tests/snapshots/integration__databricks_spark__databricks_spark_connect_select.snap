---
source: crates/runtime/tests/databricks_spark/mod.rs
description: "Databricks (mode: spark_connect) Integration Test Results"
---
+-------------------+----------------------------+----------------------------+--------------------------+--------------------------------+---------------------+---------------------+
| network           | isp                        | organization               | autonomous_system_number | autonomous_system_organization | mobile_country_code | mobile_network_code |
+-------------------+----------------------------+----------------------------+--------------------------+--------------------------------+---------------------+---------------------+
| 1.0.128.0/17      | TOT Public Company Limited | TOT Public Company Limited |                          |                                |                     |                     |
| 1.128.0.0/11      | Telstra Internet           | Telstra Internet           | 1221                     | Telstra Pty Ltd                |                     |                     |
| 100.43.128.0/24   | Firecold_Inc               | Firecold_Inc               |                          |                                |                     |                     |
| 100.43.129.0/26   | bernice_ulloa              | bernice_ulloa              |                          |                                |                     |                     |
| 100.43.129.128/28 | dewei_feng                 | dewei_feng                 |                          |                                |                     |                     |
| 100.43.129.144/28 | radoslav_ja?ar             | radoslav_ja?ar             |                          |                                |                     |                     |
| 100.43.129.160/28 | jakub_henderson            | jakub_henderson            |                          |                                |                     |                     |
| 100.43.129.176/28 | elisa_blom                 | elisa_blom                 |                          |                                |                     |                     |
| 100.43.129.192/29 | gertrude_burnham           | gertrude_burnham           |                          |                                |                     |                     |
| 100.43.129.200/29 | mariam_lovett              | mariam_lovett              |                          |                                |                     |                     |
+-------------------+----------------------------+----------------------------+--------------------------+--------------------------------+---------------------+---------------------+
